File: Java/Programmers/Kakao/Blind2021/blind2021_ex1.java
package Programmers.Kakao.Blind2021;



public class blind2021_ex1 {
    private static class Solution {
        public String solution(String new_id) {
            return new KakaoID(new_id)
                    .replaceToLowerCase()
                    .filter()
                    .toSingleDot()
                    .emptyCheck()
                    .rangeCheck()
                    .getResult();
        }

        private static class KakaoID {
            private String s;

            KakaoID(String s) { this.s = s; }

            private KakaoID replaceToLowerCase() {
                s = s.toLowerCase();
                return this;
            }

            private KakaoID filter() {
                s = s.replaceAll("[^a-z0-9._-]", "");
                return this;
            }

            private KakaoID toSingleDot() {
                s = s.replaceAll("[.]{2,}", ".")
                        .replaceAll("^[.]|[.]$", "");

                return this;
            }

            private KakaoID emptyCheck() {
                s = s.isEmpty() ? "a" : s;
                return this;
            }

            private KakaoID rangeCheck() {
                if (s.length() > 15) s = s.substring(0, 15);
                s = s.replaceAll("[.]$", "");

                StringBuilder sb = new StringBuilder(s);
                while (sb.length() <= 2)
                    sb.append(sb.charAt(s.length() - 1));

                s = sb.toString();
                return this;
            }

            private String getResult() { return this.s; }
        }
    }

    public static void main(String[] args) {
        String test[] = {
                "...!@BaT#*..y.abcdefghijklm",
                "z-+.^.",
                "=.=",
                "123_.def",
                "abcdefghijklmn.p"
        };

        for (String s : test)
            System.out.println(new Solution().solution(s));
    }
}

File: Kotlin/src/main/kotlin/programmers/kakao_blind/Blind2021_ex1.kt
package programmers.kakao_blind


private class Blind2021_ex1 {
    fun String.checkStr(): String {
        if (this.isEmpty()) return "a"
        else if (this.length > 15) {
            val str = this.substring(0 until 15)
            return if (str.endsWith('.')) str.substring(0 until 14)
            else str
        }
        return this
    }

    /*
    fun String.checkSize(): String {
        val sb = StringBuilder(this)
        while (sb.length < 3)
            sb.append(this.last())

        return sb.toString()
    }
    */

    fun solution(new_id: String): String {
        return new_id
            .map { it.lowercaseChar() }
            .filter { it.isDigit() || it.isLowerCase() || it in "-_." }
            .joinToString("")
            .replace("[.]*[.]".toRegex(), ".")
            .removePrefix(".")
            .removeSuffix(".")
            .checkStr()
            .let {
                StringBuilder(it).run {
                    while (length < 3) append(it.last())
                    toString()
                }
            }
    }
}

fun blind2021_ex1() {
    val id1 = "...!@BaT#*..y.a123bcdefghijklm"
    val id2 = "z-+.^."
    val id3 = "123_.def"
    val id4 = "abcdefghijklmn.p"
    val test = listOf(id1, id2, id3, id4)

    for (id in test) println(Blind2021_ex1().solution(id))
}